### PR TITLE
Add test for regression between 3.1.8 -> 3.1.38 where overriding dlopen() with a gracefully failing version broke

### PR DESCRIPTION
**Pull request by juj**
In #16790 the consensus was to not add a built-in build option to make `dlopen()` gracefully fail, but that users should override the pertinent JS function symbols themselves.

That is what we did at Unity with 3.1.8. Now updating to 3.1.38, I see that this override capability has regressed since some of the JS function names have changed.

This PR adds a new test for this regression between 3.1.8 -> 3.1.38 where overriding dlopen() with a gracefully failing version broke, so that user code overriding dlopen() cannot regress again. It was possibly regressed by #18638.

This way the code cannot be changed around this area without test runner being aware that it will regress end users.

### CODE REVIEW

File: src/library_dylink.js [hunk 1]
```diff
@@ -299,7 +299,7 @@ var LibraryDylink = {
   _emscripten_dlopen_js: function(handle, onsuccess, onerror, user_data) {
     abort(dlopenMissingError);
   },
-  _dlsym_js: function(handle, symbol) {
+  _dlsym_js: function(handle, symbol, symbolIndex) {
```
Comment (juj) [inline src/library_dylink.js:302]: This is a non-functional drive-by signature fix for consistency. Looks like a new parameter was added, but was not updated here.

**Comment by juj** [comment]
CI instability:

```
--2023-05-22 11:09:56--  https://nodejs.org/download/v8-canary/v20.0.0-v8-canary202302081604228b65/node-v20.0.0-v8-canary202302081604228b65-linux-x64.tar.xz
Resolving nodejs.org (nodejs.org)... 104.20.22.46, 104.20.23.46, 2606:4700:10::6814:172e, ...
Connecting to nodejs.org (nodejs.org)|104.20.22.46|:443... connected.
HTTP request sent, awaiting response... 500 Internal Server Error
2023-05-22 11:09:56 ERROR 500: Internal Server Error.
```

File: test/other/test_dlopen_graceful_fail.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
+#if !MAIN_MODULE
```
Comment (sbc100) [inline test/other/test_dlopen_graceful_fail.js:1]: `if !MAIN_MODULE && !SIDE_MODULE` ?

File: test/other/test_dlopen_graceful_fail.out [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Attempting to dlopen
+warning: dlopen: Unable to open DLL! Dynamic linking is not supported in WebAssembly builds due to limitations to performance and code size. Please statically link in the needed libraries.
+dlopen completed: ptr 0. error: (null)
+dlclose(0) completed.
+dlopen failed gracefully!
```
Comment (sbc100) [inline test/other/test_dlopen_graceful_fail.out:5]: Newline at EOF

File: test/other/test_dlopen_graceful_fail.c [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+// Some users want to instead have dlopen() perform a graceful "return 0" failure
+// where they handle the dlopen failures in code, so they want to override
+// Emscripten's built-in JS library functions.
+
+// This test verifies that the ability to override dlopen() to not hard abort
+// will not regress.
+#include <dlfcn.h>
+#include <stdio.h>
+
+int main()
```
Comment (sbc100) [inline test/other/test_dlopen_graceful_fail.c:11]: We use the LLVM style which is opening brace on same line as function decl.    I know its not consistent, especially in tests, but we are trying to tend in the right direction.

**Review by sbc100 — COMMENTED** [review]
lgtm % comments

**Review by sbc100 — COMMENTED** [review]
Does this mean we are to consider these JS functions names part of our official API and never change/add/remove them?   I'm not sure I want to do that, especially since I'm actively working on dyanmic linking. 

Perhaps at least adding this test will allow us to know when we need to update the ChangeLog to document this?

**Comment by juj** [comment]
> Does this mean we are to consider these JS functions names part of our official API and never change/add/remove them? I'm not sure I want to do that, especially since I'm actively working on dyanmic linking.

Well, we do want one or the other: either
a) make this stable so that we can rely on it, or
b) provide a build option e.g. -sABORTING_DLOPEN=0 to enable the use case out of the box.

In the linked #16790 the option b) was rejected, so that would only leave a)?

We do definitely need a way to have a graceful dlopen() one way or the other that is stable code, so we don't need to keep this on the running checklist whenever we update Emscripten in our codebase.

In the past two Emscripten updates, we have had regressions on this area, so this is an attempt to have this code stick in the repository so that the next Emscripten update will not make it a third.

I think a -sABORTING_DLOPEN=0 would be the better way forward out of the above two options? Maybe it might make sense to reconsider #16790?

**Comment by sbc100** [comment]
> > Does this mean we are to consider these JS functions names part of our official API and never change/add/remove them? I'm not sure I want to do that, especially since I'm actively working on dyanmic linking.
> 
> Well, we do want one or the other: either a) make this stable so that we can rely on it, or b) provide a build option e.g. -sABORTING_DLOPEN=0 to enable the use case out of the box.
> 
> In the linked #16790 the option b) was rejected, so that would only leave a)?
> 
> We do definitely need a way to have a graceful dlopen() one way or the other that is stable code, so we don't need to keep this on the running checklist whenever we update Emscripten in our codebase.
> 
> In the past two Emscripten updates, we have had regressions on this area, so this is an attempt to have this code stick in the repository so that the next Emscripten update will not make it a third.
> 
> I think a -sABORTING_DLOPEN=0 would be the better way forward out of the above two options? Maybe it might make sense to reconsider #16790?

I think perhaps you are right..   another alternative is to simple remove the aborting versions (i.e. maybe `-sABORTING_DLOPEN=0` the default/only option).

The other option is to use the existing `-sALLOW_UNIMPLEMENTED_SYSCALLS` setting and have dlopen work just like all other system functions (i.e. stubs exist with `-sALLOW_UNIMPLEMENTED_SYSCALLS` but link error when this is not set).

**Comment by sbc100** [comment]
Feel free to re-open if you don't like the solution I landed.

**Comment by juj** [comment]
Thanks, that will certainly do the trick for us. Not sure if there might still be room for a -sABORTING_DLOPEN=1 flag in the future, but I guess we'll see how common such a debugging need will be.
